### PR TITLE
Permission control service for saved objects

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

1. Create a standalone permission check service inside saved objects directory, as the permission control service is only related to saved objects.
2. Define some methods the service will provides.
3. Register two http routes for browser to request. 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/saved_objects/permission_control/routes/list.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import { IRouter } from '../../../http';
+import { SavedObjectsPermissionControlContract } from '../client';
+
+export const registerListRoute = (
+  router: IRouter,
+  permissionControl: SavedObjectsPermissionControlContract
+) => {
+  router.post(
+    {
+      path: '/list',
```
Comment (Hailong-am) [inline src/core/server/saved_objects/permission_control/routes/list.ts:16]: how about rename to `principles` as the implementation is about get principles of saved objects?

File: src/core/server/saved_objects/permission_control/client.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { OpenSearchDashboardsRequest } from '../../http';
+import { SavedObjectsServiceStart } from '../saved_objects_service';
+import { SavedObjectsBulkGetObject } from '../service';
+
+export enum PermissionMode {
```
Comment (wanglam) [inline src/core/server/saved_objects/permission_control/client.ts:10]: Could we move this `PermissionMode` to some common folder? So we can use it even in `public` folder.

**Review by ruanyl — APPROVED** [review]

File: src/core/server/saved_objects/permission_control/routes/list.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/routes/list.ts:16]: Agree.

File: src/core/server/saved_objects/permission_control/client.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  LibraryRead = 'library_read',
+  LibraryWrite = 'library_write',
+}
+
+export type SavedObjectsPermissionControlContract = Pick<
+  SavedObjectsPermissionControl,
+  keyof SavedObjectsPermissionControl
+>;
+
+export type SavedObjectsPermissionModes = PermissionMode | PermissionMode[] | string[];
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/permission_control/client.ts]: Can not agree more, let's make it as simple as possible in the beginning.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 4]
```diff
@@ -137,7 +163,7 @@ export class WorkspaceSavedObjectsClientWrapper {
         await this.validateAtLeastOnePermittedWorkspaces(
           object.workspaces,
           wrapperOptions.request,
-          WorkspacePermissionMode.Read
+          PermissionMode.Read
```
Comment (ruanyl) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:166]: I think we can simply pass `[PermissionMode.Read]`

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 5]
```diff
@@ -150,17 +176,20 @@ export class WorkspaceSavedObjectsClientWrapper {
...
             )
         );
       } else {
         options.workspaces = [
           'public',
           ...(await this.permissionControl.getPermittedWorkspaceIds(
-            WorkspacePermissionMode.Read,
-            wrapperOptions.request
+            wrapperOptions.request,
+            this.formatPermissionModeToStringArray(PermissionMode.Read)
```
Comment (ruanyl) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:192]: I feel simply pass `[PermissionMode.Read]` is good enough.

**Review by wanglam — APPROVED** [review]

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 5 above)
Comment (SuZhou-Joe) [reply src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:192]: Agree.

**Pull request merged**